lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,239 @@
+        rtapp_profile = cls.get_rtapp_profile(plat_info)
+        cgroup_config = cls.get_cgroup_configuration(plat_info, boost, prefer_idle)
+        cls._run_rtapp(target, res_dir, rtapp_profile, ftrace_coll, cgroup_config)
+
+        return cls(res_dir, plat_info, boost, prefer_idle)
+
+class SchedTuneBase(TestBundle):
+    """
+    Abstract class enabling the aggregation of ``SchedTuneItemBase``
+
+    :param test_bundles: a list of test bundles generated by
+        multiple ``SchedTuneItemBase`` instances
+    :type test_bundles: list
+    """
+    def __init__(self, res_dir, plat_info, test_bundles):
+        super().__init__(res_dir, plat_info)
+
+        self.test_bundles = test_bundles
+
+    @classmethod
+    def from_target(cls, target:Target, res_dir:ArtifactPath=None,
+            ftrace_coll:FtraceCollector=None) -> 'SchedTuneBase':
+        """
+        XXX: Warning if nothing is documented here ...
+        """
+        return super().from_target(target, res_dir, ftrace_coll=ftrace_coll)
+
+    @classmethod
+    def _from_target(cls, target, res_dir, ftrace_coll):
+        return cls(res_dir, target.plat_info,
+            list(cls._create_test_bundles(target, res_dir, ftrace_coll))
+        )
+
+    @classmethod
+    @abc.abstractmethod
+    def _create_test_bundles(cls, target, res_dir, ftrace_coll):
+        """
+        Collects and returns a ResultBundle per test item.
+        """
+        pass
+
+    @classmethod
+    def _create_test_bundle(cls, target, res_dir, ftrace_coll, item_cls, boost, pi):
+        """
+        Creates and returns a TestBundle for a given item class, and a given
+        schedtune configuration
+        """
+        item_dir = os.path.join(res_dir, 'boost_{}'.format(boost))
+        os.makedirs(item_dir)
+
+        logger = cls.get_logger()
+        logger.info('Running {} with boost={}, prefer_idle={}'.format(
+                    item_cls.__name__, boost, pi))
+        return item_cls.from_target(target, boost, pi, res_dir=item_dir, ftrace_coll=ftrace_coll)
+
+    def _merge_res_bundles(self, res_bundles):
+        """
+        Merge a set of result bundles
+        """
+        overall_bundle = ResultBundle.from_bool(all(res_bundles.values()))
+        for name, bundle in res_bundles.items():
+            overall_bundle.add_metric(name, bundle.metrics)
+
+        overall_bundle.add_metric('failed', [
+            name for name, bundle in res_bundles.items()
+            if bundle.result is Result.FAILED
+        ])
+        return overall_bundle
+
+class SchedTuneFreqItem(SchedTuneItemBase):
+    """
+    Runs a tiny RT rtapp task pinned to a big CPU at a given boost level and
+    checks the frequency selection was performed accordingly.
+    """
+
+    ftrace_conf = FtraceConf({
+        "events" : ["sched_switch", "cpu_frequency"]
+    }, __qualname__)
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        cpu = plat_info['capacity-classes'][-1][0]
```

Comment:
Can we grant `capacity-classes` is always ordered?
Otherwise, something like:
```python
big_cpus = [cpu for cpu,cap
            in target.plat_info['cpu-capacities'].items()
            if cap==1024]
big_cpu = big_cpus[-1]
```
Should give your the topmost big cpu.

Reply:
Yeah `capacity-classes` is always sorted, we already rely on that all over